<!DOCTYPE html>
<html>
<head><title>R: ShinyWGD</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body><div class="container">
<h1> ShinyWGD
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;ShinyWGD&rsquo; version 0.1.0</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
</ul>

<h2>Help Pages</h2>


<table style="width: 100%;">
<tr><td style="width: 25%;"><a href="analysisEachCluster.html">analysisEachCluster</a></td>
<td>Perform synteny analysis for identified clusters</td></tr>
<tr><td style="width: 25%;"><a href="bootStrapPeaks.html">bootStrapPeaks</a></td>
<td>Bootstrap Peaks in the Ks Distribution</td></tr>
<tr><td style="width: 25%;"><a href="calculateKsDistribution4wgd_multiple.html">calculateKsDistribution4wgd_multiple</a></td>
<td>Calculate the Ks Distribution for Multiple Speices</td></tr>
<tr><td style="width: 25%;"><a href="CalHomoConcentration.html">CalHomoConcentration</a></td>
<td>Compute the -log10 of Poisson Distribution</td></tr>
<tr><td style="width: 25%;"><a href="CalPvalue.html">CalPvalue</a></td>
<td>Compute the P-value of a Cluster using the Poisson Distribution</td></tr>
<tr><td style="width: 25%;"><a href="check_gff_from_file.html">check_gff_from_file</a></td>
<td>Check and Process GFF Input File from a Specific Path</td></tr>
<tr><td style="width: 25%;"><a href="check_gff_input.html">check_gff_input</a></td>
<td>Check and Prepare GFF/GTF Input File</td></tr>
<tr><td style="width: 25%;"><a href="check_proteome_from_file.html">check_proteome_from_file</a></td>
<td>Check and Process Proteome Input File From a Special Path</td></tr>
<tr><td style="width: 25%;"><a href="check_proteome_input.html">check_proteome_input</a></td>
<td>Check and Process Proteome Input File</td></tr>
<tr><td style="width: 25%;"><a href="cluster_synteny.html">cluster_synteny</a></td>
<td>Cluster Synteny Data and Generate Trees</td></tr>
<tr><td style="width: 25%;"><a href="computing_depth.html">computing_depth</a></td>
<td>Compute the Depth of Anchored Points</td></tr>
<tr><td style="width: 25%;"><a href="computing_depth_paranome.html">computing_depth_paranome</a></td>
<td>Compute the Depth of Anchored Points in a Paranome Comparison</td></tr>
<tr><td style="width: 25%;"><a href="convert_wgd2kevins.html">convert_wgd2kevins</a></td>
<td>Convert</td></tr>
<tr><td style="width: 25%;"><a href="CountOrthologs.html">CountOrthologs</a></td>
<td>Count Ortholog Genes in a Species</td></tr>
<tr><td style="width: 25%;"><a href="create_ksrates_cmd.html">create_ksrates_cmd</a></td>
<td>Create Ksrates Command Files from Shiny Input</td></tr>
<tr><td style="width: 25%;"><a href="create_ksrates_cmd_from_table.html">create_ksrates_cmd_from_table</a></td>
<td>Create Ksrates Command Files from Data Table</td></tr>
<tr><td style="width: 25%;"><a href="create_ksrates_configure_file_based_on_table.html">create_ksrates_configure_file_based_on_table</a></td>
<td>Create Ksrates Configuration File Based on Data Table</td></tr>
<tr><td style="width: 25%;"><a href="create_ksrates_configure_file_v2.html">create_ksrates_configure_file_v2</a></td>
<td>Create Ksrates Configuration File</td></tr>
<tr><td style="width: 25%;"><a href="create_ksrates_expert_parameter_file.html">create_ksrates_expert_parameter_file</a></td>
<td>Create ksrates Expert Parameter File</td></tr>
<tr><td style="width: 25%;"><a href="dfltBWrange.html">dfltBWrange</a></td>
<td>dfltBWrange</td></tr>
<tr><td style="width: 25%;"><a href="dfltCounts.html">dfltCounts</a></td>
<td>dfltCounts</td></tr>
<tr><td style="width: 25%;"><a href="downloadButton_custom.html">downloadButton_custom</a></td>
<td>Creating a Custom Download Button</td></tr>
<tr><td style="width: 25%;"><a href="drvkde.html">drvkde</a></td>
<td>drvkde</td></tr>
<tr><td style="width: 25%;"><a href="extractCluster.html">extractCluster</a></td>
<td>Extract clusters based on specified scaffolds</td></tr>
<tr><td style="width: 25%;"><a href="find_peaks.html">find_peaks</a></td>
<td>Find Peaks in a Numeric Vector</td></tr>
<tr><td style="width: 25%;"><a href="generateKsDistribution.html">generateKsDistribution</a></td>
<td>Generate the Ks Distribution</td></tr>
<tr><td style="width: 25%;"><a href="generate_ksd.html">generate_ksd</a></td>
<td>Convert</td></tr>
<tr><td style="width: 25%;"><a href="get_segments.html">get_segments</a></td>
<td>Get Segmented Data from Anchorpoints and Ks Values</td></tr>
<tr><td style="width: 25%;"><a href="is.ksv.html">is.ksv</a></td>
<td>ksv class</td></tr>
<tr><td style="width: 25%;"><a href="is.not.null.html">is.not.null</a></td>
<td>Check if an Object is Not NULL</td></tr>
<tr><td style="width: 25%;"><a href="ks_mclust_v2.html">ks_mclust_v2</a></td>
<td>ks_mclust_v2</td></tr>
<tr><td style="width: 25%;"><a href="map_informal_name_to_latin_name.html">map_informal_name_to_latin_name</a></td>
<td>Map Informal Names to Latin Names</td></tr>
<tr><td style="width: 25%;"><a href="mix_logNormal_Ks.html">mix_logNormal_Ks</a></td>
<td>Log-Normal mixturing analyses of a Ks distributions for the whole paranome</td></tr>
<tr><td style="width: 25%;"><a href="modeFinder.html">modeFinder</a></td>
<td>modeFinder</td></tr>
<tr><td style="width: 25%;"><a href="obtain_chromosome_length.html">obtain_chromosome_length</a></td>
<td>obtain_chromosome_length</td></tr>
<tr><td style="width: 25%;"><a href="obtain_chromosome_length_filter.html">obtain_chromosome_length_filter</a></td>
<td>obtain_chromosome_length_filter</td></tr>
<tr><td style="width: 25%;"><a href="obtain_coordiantes_for_anchorpoints.html">obtain_coordiantes_for_anchorpoints</a></td>
<td>Obtain coordinates for anchorpoints from GFF files</td></tr>
<tr><td style="width: 25%;"><a href="obtain_coordiantes_for_anchorpoints_ks.html">obtain_coordiantes_for_anchorpoints_ks</a></td>
<td>Obtain Coordinates and Ks Values for Anchorpoints</td></tr>
<tr><td style="width: 25%;"><a href="obtain_coordiantes_for_segments.html">obtain_coordiantes_for_segments</a></td>
<td>Obtain coordinates for segments in a comparison</td></tr>
<tr><td style="width: 25%;"><a href="obtain_coordinates_for_segments_multiple.html">obtain_coordinates_for_segments_multiple</a></td>
<td>Obtain Coordinates for Segments in Multiple Synteny Blocks</td></tr>
<tr><td style="width: 25%;"><a href="obtain_mean_ks_for_each_multiplicon.html">obtain_mean_ks_for_each_multiplicon</a></td>
<td>Compute the Mean of Ks values for Each Multiplicon</td></tr>
<tr><td style="width: 25%;"><a href="parse_EMMIX.html">parse_EMMIX</a></td>
<td>Read the EMMIX output for a range of components</td></tr>
<tr><td style="width: 25%;"><a href="parse_one_EMMIX.html">parse_one_EMMIX</a></td>
<td>Read the EMMIX output for a specify number of components</td></tr>
<tr><td style="width: 25%;"><a href="PeaksInKsDistributionValues.html">PeaksInKsDistributionValues</a></td>
<td>Find Peaks in the Ks Distribution</td></tr>
<tr><td style="width: 25%;"><a href="plot.ksv.html">plot.ksv</a></td>
<td>Draw Ks distribution using output from 'wgd'</td></tr>
<tr><td style="width: 25%;"><a href="plot_ksv_density.html">plot_ksv_density</a></td>
<td>Draw a density plot of ksv</td></tr>
<tr><td style="width: 25%;"><a href="plot_Ks_mix.html">plot_Ks_mix</a></td>
<td>Draw a Ks distribution with fitted Gaussian</td></tr>
<tr><td style="width: 25%;"><a href="read.dating_anchors.html">read.dating_anchors</a></td>
<td>Read the output of Ks values for anchor pairs from the dating pipeline</td></tr>
<tr><td style="width: 25%;"><a href="read.wgd_ksd.html">read.wgd_ksd</a></td>
<td>Read the output file of wgd ksd</td></tr>
<tr><td style="width: 25%;"><a href="read_data_file.html">read_data_file</a></td>
<td>Read Data from Uploaded File</td></tr>
<tr><td style="width: 25%;"><a href="read_KTpipeline.html">read_KTpipeline</a></td>
<td>Read file from KT pipeline</td></tr>
<tr><td style="width: 25%;"><a href="relativeRate.html">relativeRate</a></td>
<td>relativeRate</td></tr>
<tr><td style="width: 25%;"><a href="replace_informal_name_to_latin_name.html">replace_informal_name_to_latin_name</a></td>
<td>Replace Informal Names with Latin Names</td></tr>
<tr><td style="width: 25%;"><a href="resampleKsDistribution.html">resampleKsDistribution</a></td>
<td>Resample a Ks Distribution</td></tr>
<tr><td style="width: 25%;"><a href="runApp.html">runApp</a></td>
<td>The main code to run ShinyWGD</td></tr>
<tr><td style="width: 25%;"><a href="run_emmix_kmeas.html">run_emmix_kmeas</a></td>
<td>A wrapper to run EM analysis of \(ln\) Ks values with k-means</td></tr>
<tr><td style="width: 25%;"><a href="SignifFeatureRegion.html">SignifFeatureRegion</a></td>
<td>SignifFeatureRegion</td></tr>
<tr><td style="width: 25%;"><a href="SiZer.html">SiZer</a></td>
<td>SiZer (Significant Zero Crossings)</td></tr>
<tr><td style="width: 25%;"><a href="symconv.ks.html">symconv.ks</a></td>
<td>symconv.ks</td></tr>
<tr><td style="width: 25%;"><a href="symconv2D.ks.html">symconv2D.ks</a></td>
<td>symconv2D.ks</td></tr>
<tr><td style="width: 25%;"><a href="symconv3D.ks.html">symconv3D.ks</a></td>
<td>symconv3D.ks</td></tr>
<tr><td style="width: 25%;"><a href="symconv4D.ks.html">symconv4D.ks</a></td>
<td>symconv4D.ks</td></tr>
</table>
</div></body></html>
