---
title: "An Introduction to **`shinyWGD`**"
author: "Jia Li"
date: "`r Sys.Date()`"
show_toc: yes
githubEditURL: https://github.com/li081766/shinyWGD/blob/main/vignettes/intro_to_shinywgd.Rmd
output: 
   html_document:
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 3
    number_sections: true
vignette: >
  %\VignetteIndexEntry{An Introduction to shinyWGD}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<style>
blockquote {
  font-size: 14px;
  background-color: #F0FFFF;
  padding: 10px;
}
</style>

Source: https://github.com/li081766/shinyWGD/blob/main/vignettes/intro_to_shinywgd.Rmd  <img src="../inst/shinyWGD/www/images/sticker_github.png" align="right" width="120" style="margin-top: -100px;">

```{r, include=FALSE}
knitr::opts_chunk$set(
  collapse=TRUE,
  comment="#>"
)
```

```{r include=FALSE}
library(fontawesome)
```

# Getting Started

This vignette provides an introduction to WGD inference using 
the **`shinyWGD`** server or package.

**`shinyWGD`** can prepare the input and command lines for [`wgd`](https://github.com/arzwa/wgd), [`ksrates`](https://github.com/VIB-PSB/ksrates), [`i-ADHoRe`](https://www.vandepeerlab.org/?q=tools/i-adhore30), [`OrthoFinder`](https://github.com/davidemms/OrthoFinder), and [`Whale`](https://github.com/arzwa/Whale.jl/tree/master) 
and then study the whole genome duplication events (WGDs) of the interesting species.

This page provides a brief introduction to this software. 

<br></br>
<br></br>

# Installation

## Install from CRAN

```r
install.packages("shinyWGD")
```

## Install the latest development version from GitHub

```r
if (!require("devtools")) {
  install.packages("devtools")
}
devtools::install_github("li081766/shinyWGD", dependencies=TRUE, build_vignettes=FALSE)
```

<br></br>
<br></br>

# Usage

```r
library(shinyWGD)
runshinyWGD()
help(package="shinyWGD")
```

## Data Preparation

### <font color="#23B395">Preparing the input for `wgd`, `ksrates`, `i-ADHoRe` or `OrthoFinder`</font>

Please select the proper number of species you want to study in <font color="9f5000">Select Number of Species to Analyze:</font> first.

  <blockquote>
  `r fa(name="exclamation-triangle", fill="green")` If only one species is set, the core program will be set to <b>`wgd`</b>, otherwise, <b>`ksrates`</b> will be chosen.</blockquote>

**`shinyWGD`** provides two options to upload the data in the `Data Preparation` Page.

#### <i>Option One</i>

Users can prepare a **Tab-Separated** file in which each row includes the data of a species. 

This option is suitable to study multiple species. 

The detailed format of this file is below: </br>

  + **`Tab-Separated file`**
  
    <table style="background-color: #FFFAF0; border: 1px solid #000; border-collapse: collapse;">
      <tr>
        <th style="padding: 5px;"><i>latin name</i></th>
        <th style="padding-left: 50px; padding: 5px;">cds path<sup>a</sup></th>
        <th style="padding-left: 40px; padding: 5px;">gff path<sup>b</sup> (optional)</th>
      </tr>
      <tr>
        <td style="padding: 5px;"><i>Elaeis guineensis</i></td>
        <td style="padding-left: 50px; padding: 5px;">PATH/elaeis.fasta</td>
        <td style="padding-left: 40px; padding: 5px;">PATH/elaeis.gff</td>
      </tr>
      <tr>
        <td style="padding: 5px;"><i>Oryza sativa</i></td>
        <td style="padding-left: 50px; padding: 5px;">PATH/aryza.fasta</td>
        <td style="padding-left: 40px; padding: 5px;">PATH/oryza.gff3</td>
      </tr>
      <tr>
        <td style="padding: 5px;"><i>Asparagus officinalis</i></td>
        <td style="padding-left: 50px; padding: 5px;">PATH/asparagus.fasta</td>
        <td style="padding-left: 40px; padding: 5px;"></td>
      </tr>
    </table>

    - **<sup>a</sup>** `cds` is the coding region of a gene or the portion of a gene's DNA or RNA that codes for protein. This file can be `fasta` or `gzipped-fasta`. <br>This column is **mandatory**. 
    - `r fa(name="exclamation-triangle", fill="red")` Do not contain the [alternative isoforms](https://en.wikipedia.org/wiki/Alternative_splicing) of each gene.
    - **<sup>b</sup>** `gff` is the format consists of one line per feature, each containing nine columns of data, plus optional track definition lines, see more [click here](https://www.ensembl.org/info/website/upload/gff.html). <br>The file can be `gff`, `gff3`, `gzipped-gff`, or `gzipped-gff3`. <br>This column is **mandatory** for **`focal species`** and is **optional** for **`other species`**.</br>

    - **`focal species`** is the parameter in `ksrates`, which will be set in the [`configure file`](https://ksrates.readthedocs.io/en/latest/configuration.html) of `ksrates`.

After successfully uploading the tab-separated file, the number of species in the file will be calculated. 

  - If only one species needed to be studied, **`shinyWGD`**
   will generate the command line for `wgd` by clicking the <font style="background-color: #548C00; color: white; padding: 2px; border-radius: 2px;">`r fa(name="code")` Create WGD Codes</font> or <font style="background-color: #548C00; color: white; padding: 2px; border-radius: 2px;">`r fa(name="code")` Create i-ADHoRe Codes</font> button in the right **`Setting`** panel. 
    
    - Users can check the codes by clicking the <font color="#ad5a5a">`r fa(name="share")` Check WGD Codes</font> or <font color="#ad5a5a">`r fa(name="share")` Check i-ADHoRe Codes</font> button.
    
    - Then, users can click the <font style="background-color: #508ab2; color: white; padding: 2px; border-radius: 2px;">`r fa(name="download")` Download Analysis Data</font> button to download the prepared files and script for `wgd` program. 

  - If multiple species needed to be study, **`shinyWGD`**
   will generate the command line for `ksrates`. 
    
    - Two more parameters should be added.
      - [`focal species`](https://ksrates.readthedocs.io/en/latest/configuration.html) set the focal species for `ksrates`.
      - `newick tree` upload a tree file in `newick` format. The tree should contain all the species.
    
    - Then, users can click the <font style="background-color: #548C00; color: white; padding: 2px; border-radius: 2px;">`r fa(name="code")` Create Ksrates Codes</font> button to generate the `configure file` and command line for `ksrates`.
    
    - After that, users can click the <font color="#ad5a5a">`r fa(name="share")` Check Ksrates Codes</font> button to review the `configure file` and command line for `ksrates`.
    
    - Then, users can click the rest buttons to generate the codes for `i-ADHoRe` and `OrthoFinder`.
    
    - Finally, users can click the <font style="background-color: #508ab2; color: white; padding: 2px; border-radius: 2px;">`r fa(name="download")` Download Analysis Data</font> button to download the prepared files and script for `ksrates`, `i-ADHoRe`, and `OrthoFinder` program. 

#### *Option Two*

This option provide users a more straightforward way to upload data.

  - After selecting the number of studied species, the corresponding uploading panels will be created. 
  
  - Three items (`latin name`, `cds fasta`, and annotation `gff`) of each species should be set or uploaded by users. 
  
  - Once more than one species needed to be analyzed, the `focal species` and `newick tree` should be defined by users.
  
  - Then users can click the <font style="background-color: #548C00; color: white; padding: 2px; border-radius: 2px;">`r fa(name="code")` Create WGD Codes</font> or <font style="background-color: #548C00; color: white; padding: 2px; border-radius: 2px;">`r fa(name="code")` Create Ksrates Codes</font> button to create the codes for `wgd` or `ksrates` according to the species number.
  
  - After reviewing and downloading the necessary files, users can execute `wgd`, `ksrates`, `i-ADHoRe` or `OrthoFinder` in their local environment. 
  
    
 <blockquote>
  `r fa(name="exclamation-triangle", fill="red")` **`shinyWGD`** will exclude genes with internal stop codons and genes whose lengths are not divisible by three. You can find details on the gene filtering process in the *Sequence_processing.log* file located in the **`Analysis_Date`** folder.
</blockquote>

### <font color="#23B395">Executing `wgd`, `krates`, `i-ADHoRe` or `OrthoFinder`</font>

After decompressing the downloaded compressed file (**`Analysis_Date.tgz`**), users will find the `bash` and `configure` files (`run_wgd.sh` / `run_ksrates.sh`, `run_paralog_ks_rest_species.sh`, `run_iadhore.sh`, `run_orthofinder.sh`). 
  
Users can execute them in their local environment.  
    
- How to install `wgd` to your local environment, please read the [`wgd` document](https://github.com/arzwa/wgd).
- How to install `ksrate` to your local environment, please read the [`ksrates` document](https://ksrates.readthedocs.io/en/latest/installation.html).
- How to install `i-ADHoRe` to your local environment, please read the [`i-ADHoRe` document](https://www.vandepeerlab.org/?q=tools/i-adhore30).
- How to install `OrthoFinder` to your local environment, please read the [`OrthoFinder` document](https://github.com/davidemms/OrthoFinder).
    
<blockquote>
  `r fa(name="exclamation-triangle", fill="green")` When using <font color="green">`i-ADHoRe`</font> to do synteny analysis, the alignment will be produced. The alignment process will use the <font color="green">[`diamond`](https://github.com/bbuchfink/diamond)</font> program. So please install <font color="green">[`diamond`](https://github.com/bbuchfink/diamond)</font> program first.
</blockquote>

<blockquote>
  `r fa(name="exclamation-triangle", fill="green")` To compute the <i>K</i><sub>s</sub> value of each anchored gene pair and the <i>K</i><sub>s</sub> unit tree of studied species, the [`MAFFT`](https://mafft.cbrc.jp/alignment/software/), [`trimAl`](http://trimal.cgenomics.org/), and [`PAML`](http://abacus.gene.ucl.ac.uk/software/paml.html) will be used for aligning the proteins, obtaining the nucleotide alignment and calculating the <i>K</i><sub>s</sub> value, respectively. So please install <font color="green">[`MAFFT`](https://mafft.cbrc.jp/alignment/software/)</font>, <font color="green">[`trimAl`](http://trimal.cgenomics.org/)</font>, and <font color="green">[`PAML`](http://abacus.gene.ucl.ac.uk/software/paml.html)</font> first.
  
  
</blockquote>

### <font color="#23B395">Extracting the tree from [TimeTree.org](http://www.timetree.org/)</font>

  If users don’t ensure the evolutionary relationships among the studied species, **`shinyWGD`**
   provides an alternative way to generate the newick tree through interacting with [TimeTree.org](http://www.timetree.org/) database.

  <blockquote>
  `r fa(name="exclamation-triangle", fill="red")` TimeTree database does not include all the species.
  <br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;If the species are not included in the TimeTree database, you should ensure the relationship in other ways.</blockquote>

Please click the <font style="background-color: #548C00; color: white; padding: 2px; border-radius: 2px;">`r fa(name="tree")` Extract Tree</font> button to extract a tree from TimTree.org database.<br>
  
After successfully obtaining the tree, users can review the tree in the box below <font style="background-color: #548C00; color: white; padding: 2px; border-radius: 2px;">`r fa(name="tree")` Extract Tree</font> button.

<br><img src="../inst/shinyWGD/www/images/timetree.png" alt="" class="docImage" width="70%"></br>

Users need to save this tree to a local file and then upload the file to `newick tree` panel in the <b>`Setting`</b> block.

## <img src="../inst/shinyWGD/www/images/ksIcon.svg" alt="Icon" width="20" height="20">  <i>K</i><sub>s</sub> Age Distribution Analysis

After successfully running the corresponding shell files locally, the necessary outputs will be saved in the **`Analysis_Date`** folder. Then users can choose this folder in the `Uploading` panel in <img src="../inst/shinyWGD/www/images/ksIcon.svg" alt="Icon" width="10" height="10"> page.

  <blockquote>
  `r fa(name="exclamation-triangle", fill="red")` Users should choose the folder which is originated from the folder created by **`shinyWGD`** program to avoid the failure of displaying the correct file names.</blockquote>

### <font color="#23B395">Paralog <i>K</i><sub>s</sub> age distribution</font>

Once selecting one / multiple species from the drop-down menu of <font style="background-color: #17a2b8; color: #EBD1FC; padding: 2px; border-radius: 2px;">`r fa(name="list")` Paralog <i>K</i><sub>s</sub> &#x25BC; </font>, users can click the <font style="background-color: #548C00; color: white; padding: 2px; border-radius: 2px;">`r fa(name="cog")` Configure <b><i>K</i><sub>s</sub></b> Analysis</font> to create the output panel in the right side of the page. 

After clicking the <font style="background-color: #548C00; color: white; padding: 2px; border-radius: 2px;">`r fa(name="play")` Start <b><i>K</i><sub>s</sub></b> Analysis</font>, the <i>K</i><sub>s</sub> age distribution of the selected species will be generated. 

**`shinyWGD`** employs two methods to model the potential peaks of the distribution. 

First, **`shinyWGD`** fits normal mixture models to the <b><i>K</i><sub>s</sub></b> data using maximum likelihood with EMMIX. **`shinyWGD`** will test one to ten components with 500 random starts. Fitted models are selected by the Bayesian Information Criterion (BIC) within EMMIX. Finally, the fitted models are drawn above the bar plot of each species.

Second, **`shinyWGD`** uses SiZer to identify significant peaks based on the significant zero crossings of derivatives from kernel density estimation. Then, a subplot to show the output of SiZer will be added under the bar plot.

  <blockquote>
Please refer to the papers below to see details about EMMIX and SiZer.
  
  - McLachlan G, Peel D. (1999) <i>The EMMIX algorithm for the fitting of normal and t-components</i>. <b>Journal of Statistical Software</b>. 4:1–14.
  - Barker, M.S., Kane, N.C., Matvienko, M., Kozik, A., Michelmore, R.W., Knapp, S.J. and Rieseberg, L.H., (2008) <i>Multiple paleopolyploidizations during the evolution of the Compositae reveal parallel patterns of duplicate gene retention after millions of years</i>. <b>Molecular biology and evolution</b>, 25(11), pp.2445-2455.
  - Chaudhuri, P. and Marron, J.S., (1999) <i>SiZer for exploration of structures in curves</i>. <b>Journal of the American Statistical Association</b>, 94(447), pp.807-823.
</blockquote>

The example output is below.

<img src="../inst/shinyWGD/www/images/Paralogous_Multiple_Species_Ks-01.jpg" width="90%">

### <font color="#23B395">Ortholog <i>K</i><sub>s</sub> age distribution</font>

Users need to select one species from the <font color="#38B0E4">Group A</font> of the <font style="background-color: #17a2b8; color: #FF9191; padding: 2px; border-radius: 2px;">`r fa(name="list")` Ortholog <i>K</i><sub>s</sub> &#x25BC;</font> as the focal species. Then users can choose multiple species from the <font color="#B97D4B">Group B</font> to compute the density of <i>K</i><sub>s</sub> age distribution between the focal species and the species in the <font color="#B97D4B">Group B</font>. 

To obtain the potential peak of the plot, **`shinyWGD`** will fit the peak by using the Epanechnikov kernel funciton. Then 95% confidence interval (CI) is estimated by 1000 iterations.

The example output is below.

<img src="../inst/shinyWGD/www/images/Wgd_plot_ortholog.svg" width="60%">

In the above figure, the dash line is the peak of a combination of focal species and another species. The colored rectangle is the 95% CI. Users can retrieve the detail information about the density, the peaks, and the 95% CIs by hovering the mouse above the lines or rectangles.  

### <font color="#23B395">Substitution rate correction</font>

When comparing the one-to-one ortholog <i>K</i><sub>s</sub> distributions between studied species and the reference species, the different <i>K</i><sub>s</sub> peaks for the same speciation event may exist, suggesting different synonymous substitution rates among the studied species, resulting in an over- or under-estimate for the divergence between the studied species and the reference species. 

To quantify the difference in substitution rates, **`shinyWGD`** employs the relative rate tests. First, the Newick tree, which contains the phylogenetic relationship, will be used to ensure enough species. Then, users need to define a <b> reference species</b> and an <b>outgroup species</b> to determine a tree structure among the species which will be used in the relative rate test. Then, the <i>K</i><sub>s</sub> distance between two species is estimated by the mode of their ortholog KS distribution. Using the <i>K</i><sub>s</sub> distance between the reference species and the studied species, the <i>K</i><sub>s</sub> distance between the outgroup species and the reference species, the <i>K</i><sub>s</sub> distance between the outgroup species and the studied species, the distances to the studied species and the reference species after their divergence are computed. Finally, ortholog <i>K</i><sub>s</sub> between the studied species and the reference species is corrected by the double of the <i>K</i><sub>s</sub> distance to the reference species, assuming that studied species has the same substitution rate as the reference species. 

The example output is below.

<img src="../inst/shinyWGD/www/images/Rate_correction_with_paralogous_species.svg" width="60%">

In the above figure, Kernel-density estimates (KDEs) of <i>K</i><sub>s</sub> distribution for one-to-one orthologues between <i>V. vinifera</i> and other four species, namely <i>A. officinalis</i>, <i>E. guineesis</i>, <i>A. comosus</i>, and <i>Z. marina</i>. The bar and line plots are the paralog<i>K</i><sub>s</sub> distributions of <i>E. guineensis</i>. As the modes (peaks) of the KDE all represent the distances between <i>V. vinifera</i> and the compared species, the differences observed among the <i>K</i><sub>s</sub> values of the modes indicate substitution rate variations among the studied species. The segments with arrows above the KDE denote, from bottom to top, the <i>K</i><sub>s</sub> distances computed for <i>A. comosus</i> (dark red), <i>E. guineensis</i> (yellow), and <i>A. officinalis</i> (blue) using <i>Z. marina</i> (purple) as the reference and <i>V. vinifera</i> as the outgroup. The dotted lines align with the observed modes in the KDE of the ortholog <i>K</i><sub>s</sub> distributions. The black dots on the segments represent the divergence between <i>Z. marina</i> with the other three species; hence, the lengths of segments pointed to the right show the <i>K</i><sub>s</sub> accumulated in <i>A. comosus</i>, <i>E. guineensis</i> and <i>A. officinalis</i>, whereas the grey segments pointed to the left show the <i>K</i><sub>s</sub> distance between <i>V. vinifera</i> and the divergence of the other species. The grey rectangles show the 95% confidence intervals of the <i>K</i><sub>s</sub>  inferred from 200 bootstraps.

## <img src="../inst/shinyWGD/www/images/syntenyIcon.svg" alt="Icon" width="20" height="29">  Synteny Analysis

After uploading the folder of `i-ADHoRe` in this page, users can study the *intra-*, *inter-*, and *multiple-* species synteny, as well as the *clustering analysis*.

### <font color="#23B395">Intra- or Inter-species alignment</font>

After selecting the studied species, users can click the <font style="background-color: #548C00; color: white; padding: 2px; border-radius: 2px;">`r fa(name="cog")` Configure Analysis</font> to create the setting and output panel in the right side of the page. 

Users can set the threshold for the number of anchor pairs in each multiplicon to filter out the small multiplicons which contain less anchor pairs. 

The dot plot and parallel link plot will be generated.

<img src="../inst/shinyWGD/www/images/Elaeis_guineensis.self.dot_plot-01.jpg" width="60%">
<img src="../inst/shinyWGD/www/images/Elaeis_guineensis.Parallel.svg" width="80%">

### <font color="#23B395">Multiplicon-level synteny</font>

Users can also search a certain gene to check the location of the gene in the multiplicon in the <font color="green">Multiplicon-level Synteny</font> panel and then plot the searched multiplicon.

<img src="../inst/shinyWGD/www/images/microSyn_one_species.svg" width="80%">

### <font color="#23B395">Multiple species alignment</font>

In the multiple species alignment panel, users can define the order of species in the plot. Users can change the color of species by clicking the text of the species name. When the mouse hovers a link, the related links with this link among species can be highlighted using different colors. 

<img src="../inst/shinyWGD/www/images/Multiple_Species_Alignment.Parallel-01.jpg" width="80%">
  
### <font color="#23B395">Clustering analysis</font>

**`shinyWGD`** employs the Hierarchical clustering method for constructing putative ancestral regions (PARs). The segments are identified from the outputs of `i-ADHoRe`, and the **homolog concentration scores** are calculated using `-log(p)`, where `p` is the probability of the observed number of homolog pairs as modeled by a Poisson distribution.

  <blockquote>
  <font color="green">`r fa(name="exclamation-triangle")`</font> To read more about <b><i>homolog concentration scores</i></b>, please refer to Putnam NH, et al. (2008) <i>The amphioxus genome and the evolution of the chordate karyotype</i>. <b>Nature</b> 453:1064–1071.</blockquote> 
  
For each segment, the array of scores against all segments in the other genome forms a unique profile. The segments are then clustered based on the similarity of these profiles (determined by **Pearson correlation coefficient r**) using the average linkage method. The default cutoff of r is 0.3. Users can change this cutoff through the sliding block. The faint yellow rectangle will highlight the significantly enriched regions (`P < 1E-10`). Then, users can zoom in each PAR through the <font color="green"><b>Putative Ancestral Regions</b></font> block.

<img src="../inst/shinyWGD/www/images/Two_species.cluster-01.jpg" width="80%">

In the example, we used 0.5 as the cutoff for **r** and 10 to filter out the segment with fewer anchor points. In total, we identified 61 PARs. 

A single PAR zooms in is below.

<img src="../inst/shinyWGD/www/images/PAR.cluster.svg" width="50%">

## <img src="../inst/shinyWGD/www/images/ksTreeIcon.svg" alt="Icon" width="20" height="20"> Tree Building

  In this page, users can generate the <i>K</i><sub>s</sub> unit tree, time tree, and a joint tree with <i>K</i><sub>s</sub> unit tree and time tree.

### <font color="#23B395"><i>K</i><sub>s</sub> unit tree</font>

**`shinyWGD`** uses the single-copy gene families built by *OrthoFinder* to calculate the <i>K</i><sub>s</sub> unit tree. Users can check the tree in the *Orthofinder_wd* folder, named *singleCopyGene.ds_tree.newick*. Once uploading the tree file in the <font color="green"><i>K</i><sub>s</sub> Tree</font> File panel, the <i>K</i><sub>s</sub> unit tree will displayed in the right side of the page. Users can click the text of the species name to change the color or add a symbol to the species. Users can click the branch to add a <i>K</i><sub>s</sub> peak into the branch.

Users can upload a file which contains the <i>K</i><sub>s</sub> peaks of each studied species to place these peaks into the <i>K</i><sub>s</sub> unit tree.

  <!--
  + **<i>K</i><sub>s</sub> Peaks file** separated by Tab
  |<i>latin name</i>      |      <i>K</i><sub>s</sub> peak |        95% CI   | color    |
  |--------|:--------|:--------|:--------|
  |Elaeis_guineensis      |      0.25        |0.19-0.26  |  #79FF79|
  |Oryza_sativa           |      1.1         |1.02-1.3   |  #5CADAD|
  |Oryza_sativa           |      0.3         |0.22-0.4   |  #5CADAD|
  -->

  + **<i>K</i><sub>s</sub> Peaks file** separated by Tab

    <table style="background-color: #FFFAF0; border: 1px solid #000; border-collapse: collapse;"">
      <tr>
        <th style="padding: 5px;"><i>latin name</i></th>
        <th style="padding-left: 30px; padding: 5px;"><i>K</i><sub>s</sub> peak</th>
        <th style="padding-left: 40px; padding: 5px;">95% CI</th>
        <th style="padding-left: 30px; padding: 5px;">color</th>
      </tr>
      <tr>
        <td style="padding: 5px;">Elaeis_guineensis</td>
        <td style="padding-left: 30px; padding: 5px;">0.25</td>
        <td style="padding-left: 40px; padding: 5px;">0.19-0.26</td>
        <td style="padding-left: 30px; padding: 5px;">#79FF79</td>
      </tr>
      <tr>
        <td style="padding: 5px;">Oryza_sativa</td>
        <td style="padding-left: 30px; padding: 5px;">1.1</td>
        <td style="padding-left: 40px; padding: 5px;">1.02-1.3</td>
        <td style="padding-left: 30px; padding: 5px;">#5CADAD</td>
      </tr>
      <tr>
        <td style="padding: 5px;">Oryza_sativa</td>
        <td style="padding-left: 30px; padding: 5px;">0.3</td>
        <td style="padding-left: 40px; padding: 5px;">0.22-0.4</td>
        <td style="padding-left: 30px; padding: 5px;">#5CADAD</td>
      </tr>
    </table>

  <br>

  The example output is below.

  <img src="../inst/shinyWGD/www/images/ksTree.Plot.svg" width="95%">

### <font color="#23B395">Time tree</font>

Users can upload a nexus tree withe the divergence time information. The format of the tree is the output file of *MCMCTree*, named *FigTree.tre*. Be careful with the time unit, please use the 100 million yeas as the time scale. Users can click the text of the species name to change the color or add a symbol to the species. Users can click the branch to add a WGD into the branch.

Users can upload a file which contains the WGD events of each studied species to place these WGDs into the tree.

  + **WGD Events file** separated by Tab

    <table>
      <tr>
        <th><i>Latin name</i></th>
        <th style="padding-left: 50px;">WGD events<sup>a</sup></th>
        <th style="padding-left: 30px;">color</th>
      </tr>
      <tr>
        <td>Elaeis_guineensis</td>
        <td style="padding-left: 50px;">0.85,0.1-0.2</td>
        <td style="padding-left: 30px;">#79FF79</td>
      </tr>
      <tr>
        <td>Oryza_sativa</td>
        <td style="padding-left: 50px;">0.32,0.1-0.2,1.26-1.3</td>
        <td style="padding-left: 30px;">5CADAD</td>
    </table>

  - **<sup>a</sup>** WGD events can be a single time point or a time interval. Please use "**,**" to separate multiple WGD events within a species. 
  <br>

The example output is below.

<img src="../inst/shinyWGD/www/images/timeTree.Plot.svg" width="95%">

### <font color="#23B395"><i>K</i><sub>s</sub> unit tree and TimeTree joint tree</font>

If users upload both <i>K</i><sub>s</sub> unit tree and the time tree, a joint <i>K</i><sub>s</sub> unit tree and time tree will be generated. Users can also upload the <i>K</i><sub>s</sub> peak file and WGD time file to highlight the WGD events.

The example output is below.

<img src="../inst/shinyWGD/www/images/ksTree_timeTree.jointPlot.svg" width="95%">

In the above figure, the left subplot is the <i>K</i><sub>s</sub> unit tree. The right subplot is the time tree. Users can also add the <i>K</i><sub>s</sub> peaks or WGD events into the tree when clicking the branch.


## <img src="../inst/shinyWGD/www/images/treeReconciliationIcon.svg" alt="Icon" width="20" height="20"> Gene Tree – Species Tree Reconciliation Analysis

In this page, users can use `Whale` to operate the gene tree - species tree reconciliation analysis.

Users need to upload two inputs to operate `Whale.jl`. One is the species time tree in `Newick` format. Another is the folder which includes all ALE files. If users use `OrthoFinder` in **`shinyWGD`**, users can find the folder in `Orthofinder_wd`, named *all_tree_ALE_files*. Users can also follow the [instruction](https://github.com/arzwa/whaleprep/tree/master) of `Whale.jl` to prepare the ALE files by using the gene families built by other software.

Once the tree file uploaded, users can check the topology of the tree in the right panel. Then users can add the **hypothetical WGD event** to a certain branch by clicking the branch.

  <blockquote>
  `r fa(name="exclamation-triangle", fill="red")` Users should use <font color="green"><b>wgd</b></font> as the prefix for the name of <b>hypothetical WGD event</b> to satisfy the requirements of <i>Whale.jl</i>.</blockquote>

 <img src="../inst/shinyWGD/www/images/speciesTree.whale.svg" width="800px">

In the above tree figure, four hypothetical WGD events are tested, named as *wgd1*, *wgd2*, *wgd3* and *wgd4*, respectively.

Users can review the added **hypothetical WGD event** in the **<font color='#AD1F1F'>Hypothetical WGDs</font> to test** panel in the left side.

Users can define the **Base Model** in the **Base Model** for **Whale** panel and set the **Chain** rounds in the **Set the <b><font color='orange'>Chain</font></b> for <b><i>Whale</b></i>** panel.

After setting all the parameters, users can start `Whale` by clicking the <font style="background-color: #548C00; color: white; padding: 2px; border-radius: 2px;">`r fa(name="play")` Start <b><i>Whale</i></b></font> button.

After `Whale` is done, users can check the output in the right side.

The example output is below.

<img src="../inst/shinyWGD/www/images/Whale_output.jpg" width="800px">

<br></br>
We can judge whether the **hypothetical WGD event** is true or not based on the Bayes factor (*K*) to compare the likelihood of q = 0(H<sub>0</sub>) to the likelihood of q > 0(H<sub>1</sub>) using the Savage–Dickey density ratio.

  <blockquote>
  `r fa(name="exclamation-triangle", fill="red")` Please Note: the log10 Bayes factor is calculated. Thus, a Bayes factor <font color='red'><b>smaller than -2</b></font> could be considered as evidence in favor of the <i>q</i> ≠ 0 model compared to the <i>q</i> = 0 mode.</blockquote>

Besides, users should also check the **ESS** values. Generally, ESS values exceeding at least 100 is better for a good model, although short chains may be good for exploring and testing different models. If the ESS values below 100, please increase the **Chain Rounds** in the left setting panel.

Then, users can click the <font style="background-color: #548C00; color: white; padding: 2px; border-radius: 2px;">`r fa(name="sync")`</font> button to update the species tree to display the output of Whale.

The example output is below.

<img src="../inst/shinyWGD/www/images/speciesTree.updated.whale.svg" width="800px">

In the above plot, the WGD with the solid green bar is supported with retention rates significantly different from zero, while the hollow WGD bars are the ones with retention rates not different from zero.

Then users can download the plot by clicking the <font style="background-color: #5B5B00; color: white; padding: 2px; border-radius: 2px;">`r fa(name="download")`</font> button.

<br></br>
<br></br>

# Structure

- <font color="#6650C9">`r fa(name="home")` Home</font>
- <font color="#6650C9">`r fa(name="terminal")` Scripts</font>
  - <font color="orange">`r fa(name="microscope")` Data Preparation</font>
  - <font color="orange">`r fa(name="code")` Codes</font>
- <font color="#6650C9">`r fa(name="pencil-alt")` Analysis</font>
  - <font color="orange"><img src="../inst/shinyWGD/www/images/ksIcon.svg" alt="Icon" width="15" height="15"> Age Distribution Analysis</font>
  - <font color="orange"><img src="../inst/shinyWGD/www/images/syntenyIcon.svg" alt="Icon" width="15" height="15"> Synteny Analysis</font>
  - <font color="orange"><img src="../inst/shinyWGD/www/images/ksTreeIcon.svg" alt="Icon" width="15" height="15"> Tree Building</font>
  - <font color="orange"><img src="../inst/shinyWGD/www/images/treeReconciliationIcon.svg" alt="Icon" width="15" height="15"> Gene Tree – Species Tree Reconciliation Analysis</font>
- <font color="#6650C9">`r fa(name="question")` Help</font>

<br></br>
<br></br>

# Dependencies

- ## External Software
  - [`wgd`](https://github.com/arzwa/wgd)
  - [`ksrates`](https://github.com/VIB-PSB/ksrates)
  - [`i-ADHoRe`](https://www.vandepeerlab.org/?q=tools/i-adhore30)
  - [`Whale`](https://github.com/arzwa/Whale.jl/tree/master)
  - [`OrthoFinder`](https://github.com/davidemms/OrthoFinder)
  - [`diamond`](https://github.com/bbuchfink/diamond)
  - [`MAFFT`](https://mafft.cbrc.jp/alignment/software/)
  - [`trimAl`](http://trimal.cgenomics.org/)
  - [`PAML`](http://abacus.gene.ucl.ac.uk/software/paml.html)
  - [`MrBayes`](https://nbisweden.github.io/MrBayes/)
  - [`ALE`](https://github.com/ssolo/ALE)

- ## R packages
  - `{shiny}`
  - `{shinyjs}`
  - `{shinyFiles}`
  - `{shinyBS}`
  - `{shinyWidgets}`
  - `{shinyalert}`
  - `{bslib}`
  - `{bsplus}`
  - `{htmltools}`
  - `{stringi}`
  - `{tidyverse}`
  - `{vroom}`
  - `{english}`
  - `{data.table}`
  - `{argparse}`
  - `{dplyr}`
  - `{tools}`
  - `{seqinr}`
  - `{DT}`
 
<br></br>
<br></br>

# References

 - Zwaenepoel, A., and Van de Peer, Y., (2019) <i>wgd - simple command line tools for the analysis of ancient whole genome duplications</i>. <b>Bioinformatics</b>, bty915.

 - Sensalari, C., Maere, S., and Lohaus R., (2021) <i>ksrates: positioning whole-genome duplications relative to speciation events in KS distributions</i>. <b>Bioinformatics</b>, btab602.

  - Proost, S., Fostier, J., De Witte, D., Dhoedt, B., Demeester, P., Van de Peer, Y. and Vandepoele, K., (2012) <i>i-ADHoRe 3.0—fast and sensitive detection of genomic homology in extremely large data sets</i>. <b>Nucleic acids research</b>, 40(2), pp.e11-e11.
 
  - Zwaenepoel, A., and Van de Peer, Y., (2019) <i>Inference of ancient whole-genome duplications and the evolution of gene duplication and loss rates</i>. <b>Molecular biology and evolution</b>, 36(7), 1384-1404.
 
  - Emms, D.M., and Kelly, S., (2015) <i>OrthoFinder: solving fundamental biases in whole genome comparisons dramatically improves orthogroup inference accuracy</i>. <b>Genome biology</b>, 16(1), pp.1-14.

<br></br>
<br></br>

# Session info

```{r,sessioninfo}
sessionInfo()
```

