---
#title: "Introduction to ShinyWGD"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to ShinyWGD}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
## Introduction to <span style="color:red; background:linear-gradient(to right, red, yellow, green, cyan, yellow, red); -webkit-background-clip: text; -webkit-text-fill-color: transparent;">ShinyWGD</span>

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(ShinyWGD)
ShinyWGD::runApp()
```
Getting Started

This vignette provide an introduction to WGD inference using the `ShinyWGD` server or package.

`ShinyWGD` can prepare the input and command lines for [`wgd`](https://github.com/arzwa/wgd), [`ksrates`](https://github.com/VIB-PSB/ksrates), [`i-ADHoRe`](https://www.vandepeerlab.org/?q=tools/i-adhore30), [`OrthoFinder`](https://github.com/davidemms/OrthoFinder), and [`Whale`](https://github.com/arzwa/Whale.jl/tree/master) and then study the whole genome duplication events (WGDs) of the interesting species.

This page provides a brief introduction to this software. 

Please refer to the <font color='#6650C9'>[ShinyWGD server](https://bioinformatics.psb.ugent.be/webtools/ShinyWGD)</font> to see the detailed usage. 

### Installation

* Install from CRAN:

```r
install.packages("ShinyWGD")
```

* Install the latest development version from GitHub (requires [devtools](https://github.com/hadley/devtools) package):

```r
if (!require("devtools")) {
  install.packages("devtools")
}
devtools::install_github("li081766/ShinyWGD", dependencies = TRUE, build_vignettes = FALSE)
```

### The Structure of <span style="color:red; background:linear-gradient(to right, red, yellow, green, cyan, yellow, red); -webkit-background-clip: text; -webkit-text-fill-color: transparent;">ShinyWGD</span>

- ##### <font color="#6650C9"><i class="fa-solid fa-info"></i> Introduction</font>
- ##### <font color="#6650C9"><i aria-label="terminal icon" class="fa fa-terminal fa-fw fa-fade" role="presentation"></i>Scripts</font>
  - <font color="orange"><i aria-label="microscope icon" class="fa fa-microscope fa-fw fa-fade" role="presentation"></i> Data Preparation</font>
  - <font color="orange"><i aria-label="code icon" class="fa fa-code fa-fw fa-fade" role="presentation"></i> Codes</font>
- ##### <font color="#6650C9"><i class="fa-solid fa-pencil" role="presentation"></i> Analysis</font>
  - <font color="orange"><img src="images/ksIcon.svg" alt="Icon" width="20" height="20"> Age Distribution Analysis</font>
  - <font color="orange"><img src="images/syntenyIcon.svg" alt="Icon" width="20" height="20"> Synteny Analysis</font>
  - <font color="orange"><img src="images/ksTreeIcon.svg" alt="Icon" width="20" height="20"> Tree Building</font>
  - <font color="orange"><img src="images/treeReconciliationIcon.svg" alt="Icon" width="20" height="20"> Gene Tree â€“ Species Tree Reconciliation Analysis</font>
- ##### <font color="#6650C9"><i class="fa-solid fa-question"></i> Help</font>

##### Dependencies

- ###### <font color="#6650C9">External Software</font>
  - [`wgd`](https://github.com/arzwa/wgd)
  - [`ksrates`](https://github.com/VIB-PSB/ksrates)
  - [`i-ADHoRe`](https://www.vandepeerlab.org/?q=tools/i-adhore30)
  - [`Whale`](https://github.com/arzwa/Whale.jl/tree/master)
  - [`OrthoFinder`](https://github.com/davidemms/OrthoFinder)

- ###### <font color="#6650C9">R packages</font>
  - `{shiny}`
  - `{shinyjs}`
  - `{shinyFiles}`
  - `{shinyBS}`
  - `{shinyWidgets}`
  - `{shinyalert}`
  - `{bslib}`
  - `{bsplus}`
  - `{htmltools}`
  - `{stringi}`
  - `{tidyverse}`
  - `{vroom}`
  - `{english}`
  - `{data.table}`
  - `{argparse}`
  - `{dplyr}`
  - `{tools}`
  - `{seqinr}`
  - `{DT}`

